File: eduPlatformFrontend/src/types/subject.ts
export interface SubjectDto {
  id: string;
  userId: string;
  name: string;
  description: string | null;
  nameTranslations: Record<string, string> | null;
  descriptionTranslations: Record<string, string> | null;
  icon: string | null;
  color: string | null;
  isTemplate: boolean;
  templateId: string | null;
  isActive: boolean;
  isArchived: boolean;
  gradeLevel: number | null;
  topicCount: number;
  questionCount: number;
  testCount: number;
  sortOrder: number;
  createdAt: string;
  updatedAt: string;
}

export interface CreateSubjectRequest {
  name: Record<string, string>;
  description?: Record<string, string>;
  icon?: string;
  color?: string;
  gradeLevel?: number;
}

export interface UpdateSubjectRequest {
  name?: Record<string, string>;
  description?: Record<string, string>;
  icon?: string;
  color?: string;
  isActive?: boolean;
  gradeLevel?: number;
}

// PagedResponse moved to @/types/api.ts
export type { PagedResponse } from './api';

File: eduPlatformBackend/src/main/java/uz/eduplatform/modules/content/domain/Subject.java
package uz.eduplatform.modules.content.domain;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;
import uz.eduplatform.modules.auth.domain.User;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Entity
@Table(name = "subjects")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EntityListeners(AuditingEntityListener.class)
@org.hibernate.annotations.SQLRestriction("deleted_at IS NULL")
public class Subject {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id")
    private User user;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(nullable = false, columnDefinition = "jsonb")
    private Map<String, String> name;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private Map<String, String> description;

    @Column(length = 500)
    private String icon;

    @Column(length = 7)
    private String color;

    @Builder.Default
    @Column(name = "is_template")
    private Boolean isTemplate = false;

    @Column(name = "template_id")
    private UUID templateId;

    @Builder.Default
    @Column(name = "is_active")
    private Boolean isActive = true;

    @Builder.Default
    @Column(name = "is_archived")
    private Boolean isArchived = false;

    @Builder.Default
    @Column(name = "topic_count")
    private Integer topicCount = 0;

    @Builder.Default
    @Column(name = "question_count")
    private Integer questionCount = 0;

    @Builder.Default
    @Column(name = "test_count")
    private Integer testCount = 0;

    @Column(name = "grade_level")
    private Integer gradeLevel;

    @Builder.Default
    @Column(name = "sort_order")
    private Integer sortOrder = 0;

    @OneToMany(mappedBy = "subject", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("sortOrder ASC")
    @Builder.Default
    private List<Topic> topics = new ArrayList<>();

    @CreatedDate
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "deleted_at")
    private LocalDateTime deletedAt;
}
